In [1]:
import pandas as pd

df = pd.read_csv("/content/retail_sales_dataset.csv")
print("Shape:", df.shape)
df.head()

Shape: (1060, 10)


,Order_ID,Order_Date,Customer_ID,Category,Product,Quantity,Unit_Price,Total_Amount,City,Payment_Mode
0,ORD10114,2025-04-29,CUST247,Home,Table Lamp,3,99,297,Pune,UPI
1,ORD10919,2025-04-01,CUST180,Electronics,Headphones,1,999,999,Kolkata,Cash
2,ORD10492,2025-10-11,CUST376,Grocery,Cooking Oil,4,1499,5996,Mumbai,Card
3,ORD10928,2025-08-25,CUST242,Home,Bedsheet,5,499,2495,Lucknow,Card
4,ORD10603,2025-02-12,CUST149,Clothing,Jeans,3,999,2997,Delhi,Card


In [2]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1060 entries, 0 to 1059
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Order_ID      1060 non-null   object
 1   Order_Date    1060 non-null   object
 2   Customer_ID   1060 non-null   object
 3   Category      1060 non-null   object
 4   Product       1060 non-null   object
 5   Quantity      1060 non-null   int64 
 6   Unit_Price    1060 non-null   int64 
 7   Total_Amount  1060 non-null   int64 
 8   City          1060 non-null   object
 9   Payment_Mode  1060 non-null   object
dtypes: int64(3), object(7)
memory usage: 82.9+ KB


In [3]:
print("Full-row duplicates:", df.duplicated().sum())
full_dup = df[df.duplicated(keep=False)].sort_values("Order_ID")
full_dup.head(10)

Full-row duplicates: 40


,Order_ID,Order_Date,Customer_ID,Category,Product,Quantity,Unit_Price,Total_Amount,City,Payment_Mode
398,ORD10007,2025-12-05,CUST216,Clothing,Jeans,1,99,99,Kolkata,UPI
194,ORD10007,2025-12-05,CUST216,Clothing,Jeans,1,99,99,Kolkata,UPI
921,ORD10050,2025-01-19,CUST289,Electronics,Headphones,2,999,1998,Jaipur,Card
114,ORD10050,2025-01-19,CUST289,Electronics,Headphones,2,999,1998,Jaipur,Card
372,ORD10079,2025-05-20,CUST323,Home,Bedsheet,4,199,796,Kanpur,Cash
783,ORD10079,2025-05-20,CUST323,Home,Bedsheet,4,199,796,Kanpur,Cash
495,ORD10105,2025-01-28,CUST383,Electronics,Mobile Charger,3,99,297,Mumbai,Card
737,ORD10105,2025-01-28,CUST383,Electronics,Mobile Charger,3,99,297,Mumbai,Card
660,ORD10109,2025-10-16,CUST255,Electronics,Mobile Charger,2,2499,4998,Lucknow,Cash
98,ORD10109,2025-10-16,CUST255,Electronics,Mobile Charger,2,2499,4998,Lucknow,Cash


In [4]:
print("Order_ID duplicates:", df.duplicated(subset=["Order_ID"]).sum())
key_dup = df[df.duplicated(subset=["Order_ID"], keep=False)].sort_values("Order_ID")
key_dup.head(10)

Order_ID duplicates: 60


,Order_ID,Order_Date,Customer_ID,Category,Product,Quantity,Unit_Price,Total_Amount,City,Payment_Mode
398,ORD10007,2025-12-05,CUST216,Clothing,Jeans,1,99,99,Kolkata,UPI
194,ORD10007,2025-12-05,CUST216,Clothing,Jeans,1,99,99,Kolkata,UPI
477,ORD10038,2025-09-15,CUST371,Electronics,Headphones,5,499,2495,Delhi,Cash
869,ORD10038,2025-09-15,CUST371,Electronics,Headphones,5,499,2495,Jaipur,Card
921,ORD10050,2025-01-19,CUST289,Electronics,Headphones,2,999,1998,Jaipur,Card
114,ORD10050,2025-01-19,CUST289,Electronics,Headphones,2,999,1998,Jaipur,Card
381,ORD10072,2025-09-16,CUST318,Clothing,Kurta,5,2499,12495,Kolkata,Card
86,ORD10072,2025-09-16,CUST318,Clothing,Kurta,5,2499,12495,Pune,UPI
372,ORD10079,2025-05-20,CUST323,Home,Bedsheet,4,199,796,Kanpur,Cash
783,ORD10079,2025-05-20,CUST323,Home,Bedsheet,4,199,796,Kanpur,Cash


In [5]:
with pd.ExcelWriter("duplicate_report.xlsx") as w:
    full_dup.to_excel(w, sheet_name="Full_Row_Duplicates", index=False)
    key_dup.to_excel(w, sheet_name="OrderID_Duplicates", index=False)

In [6]:
clean = df.drop_duplicates()
clean = clean.drop_duplicates(subset=["Order_ID"], keep="first")
clean.to_csv("cleaned_retail_sales.csv", index=False)
print("Rows after cleaning:", len(clean))

Rows after cleaning: 1000


In [7]:
note = f"""AUDIT NOTE - Task 13: Duplicate Record Check
Original rows       : {len(df)}
Full-row duplicates : {df.duplicated().sum()}
Order_ID duplicates : {df.duplicated(subset=['Order_ID']).sum()}
Rows after cleaning : {len(clean)}
Method: drop_duplicates() on full rows, then on Order_ID (keep first).
Assumption: Order_ID should be unique per order.
"""
open("audit_note.txt", "w").write(note)
print(note)

AUDIT NOTE - Task 13: Duplicate Record Check
Original rows       : 1060
Full-row duplicates : 40
Order_ID duplicates : 60
Rows after cleaning : 1000
Method: drop_duplicates() on full rows, then on Order_ID (keep first).
Assumption: Order_ID should be unique per order.



In [8]:
from google.colab import files
for f in ["duplicate_report.xlsx", "cleaned_retail_sales.csv", "audit_note.txt"]:
    files.download(f)

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

**Conclusion**

In this task, I analyzed the Retail Sales dataset (1,060 records, 10 columns) to identify and document duplicate records using Python (Pandas).

**Two types of duplicates were found:**

Full-row duplicates (40 records): Every column value was identical to another record. These were exact copies, most likely caused by data entry or system errors, and were safely removed.
Key-column duplicates (20 records): The Order_ID was repeated, but other values such as City or Payment_Mode differed. These are conflicting records, since an Order_ID should be unique. I kept the first occurrence and documented the rest in the duplicate report so they can be verified against the source system.

After cleaning, the dataset had 1,000 unique records. This shows that duplicates can inflate sales totals and mislead analysis if not handled. Duplicate detection should be checked both on full rows and on key columns, because each method catches different problems.

**Learning outcome:** I learned how duplicated() and drop_duplicates() work, the role of the keep and subset parameters, and why documenting every cleaning step matters for a proper audit trail.